File: day_2/python/solution.py
with open("./day_2/input.txt", "r") as input:
    input = [(c.split()[0], int(c.split()[-1])) for c in input.readlines()] 


# pt1

c = [0, 0]

for action, value in input:
    if action == "up":
        c[1] -= value
    if action == "down":
        c[1] += value
    if action == "forward":
        c[0] += value

print(f"Part 1: {c[0]*c[1]}")

# pt2

c_a = [0, 0, 0]

for ac_ation, value in input:
    if ac_ation == "up":
        c_a[2] -= value
    if ac_ation == "down":
        c_a[2] += value
    if ac_ation == "forward":
        c_a[0] += value
        c_a[1] += c_a[2] * value

print(f"Part 2: {c_a[0]*c_a[1]}")

File: day_2/go/solution.go
package main

import (
	"fmt"
	"os"
	"strconv"
	"strings"
)

const (
	FORWARD = "forward"
	UP      = "up"
	DOWN    = "down"
)

type command struct {
	Action string
	Number int
}

func parse(instr string) []*command {
	var out []*command
	instr = strings.TrimSpace(instr)
	for _, c := range strings.Split(instr, "\n") {
		r := strings.Split(c, " ")
		n, _ := strconv.Atoi(r[1])
		out = append(out, &command{Action: r[0], Number: n})
	}
	return out
}

func part1(commands []*command) int {
	var h, d int

	for _, c := range commands {
		switch c.Action {
		case FORWARD:
			h += c.Number
		case DOWN:
			d += c.Number
		case UP:
			d -= c.Number
		}
	}

	return h * d
}

func part2(commands []*command) int {
	var h, d, a int

	for _, c := range commands {
		switch c.Action {
		case FORWARD:
			h += c.Number
			d += c.Number * a
		case DOWN:
			a += c.Number
		case UP:
			a -= c.Number
		}
	}

	return h * d
}

func main() {
	f, _ := os.ReadFile("./day_2/input.txt")
	input := parse(string(f))
	fmt.Println("Part 1:", part1(input))
	fmt.Println("Part 2:", part2(input))
}
